File: backend/app/services/session_service.py
"""
Business logic for session management.

SessionService handles:
- Creating and retrieving sessions
- Updating session metadata
- Deleting sessions and their associated messages
- Listing user's sessions with pagination
"""

from datetime import datetime
from zoneinfo import ZoneInfo

from sqlalchemy import desc, func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.models import ChatMessage, Session
from app.schemas.session import (
    SessionCreateRequest,
    SessionDeleteResponse,
    SessionResponse,
    SessionUpdate,
    SessionUpdateRequest,
)

KST = ZoneInfo("Asia/Seoul")


class SessionService:
    """Service for managing chat sessions."""

    @staticmethod
    async def create_session(
        db: AsyncSession,
        user_id: int,
        request: SessionCreateRequest,
    ) -> SessionResponse:
        """
        Create a new session for user.

        Args:
            db: AsyncSession database connection
            user_id: User ID who owns the session
            request: SessionCreateRequest with title and description

        Returns:
            SessionResponse with created session details
        """
        session = Session(
            user_id=user_id,
            title=request.title,
            description=request.description,
            is_active=True,
        )

        db.add(session)
        await db.flush()  # Flush to get the ID
        await db.commit()

        return SessionResponse(
            id=session.id,
            user_id=session.user_id,
            title=session.title,
            description=session.description,
            analysis_goal=session.analysis_goal,
            message_count=0,
            is_active=session.is_active,
            created_at=session.created_at.replace(tzinfo=KST),
            updated_at=session.updated_at.replace(tzinfo=KST),
        )

    @staticmethod
    async def get_session(
        db: AsyncSession,
        user_id: int,
        session_id: int,
    ) -> SessionResponse | None:
        """
        Get a specific session by ID.

        Args:
            db: AsyncSession database connection
            user_id: User ID (for authorization check)
            session_id: Session ID to retrieve

        Returns:
            SessionResponse or None if not found
        """
        # Get session with message count
        query = (
            select(Session)
            .where(
                (Session.id == session_id)
                & (Session.user_id == user_id)
            )
        )
        result = await db.execute(query)
        session = result.scalar_one_or_none()

        if not session:
            return None

        # Count messages
        count_query = select(func.count(ChatMessage.id)).where(
            ChatMessage.session_id == session_id
        )
        count_result = await db.execute(count_query)
        message_count = count_result.scalar() or 0

        return SessionResponse(
            id=session.id,
            user_id=session.user_id,
            title=session.title,
            description=session.description,
            analysis_goal=session.analysis_goal,
            message_count=message_count,
            is_active=session.is_active,
            created_at=session.created_at.replace(tzinfo=KST),
            updated_at=session.updated_at.replace(tzinfo=KST),
        )

    @staticmethod
    async def list_sessions(
        db: AsyncSession,
        user_id: int,
        limit: int = 50,
        offset: int = 0,
    ) -> tuple[list[SessionResponse], int]:
        """
        List user's sessions with pagination.

        Args:
            db: AsyncSession database connection
            user_id: User ID to filter sessions
            limit: Maximum number of sessions (default 50, max 500)
            offset: Number of sessions to skip

        Returns:
            Tuple of (SessionResponse list, total count)
        """
        # Clamp limit
        limit = min(max(limit, 1), 500)

        # Get total count
        count_query = select(func.count(Session.id)).where(
            Session.user_id == user_id
        )
        count_result = await db.execute(count_query)
        total_count = count_result.scalar() or 0

        # Get sessions ordered by created_at descending
        query = (
            select(Session)
            .where(Session.user_id == user_id)
            .order_by(desc(Session.created_at))
            .limit(limit)
            .offset(offset)
        )
        result = await db.execute(query)
        sessions = result.scalars().all()

        # Build response with message counts
        responses = []
        for session in sessions:
            count_query = select(func.count(ChatMessage.id)).where(
                ChatMessage.session_id == session.id
            )
            count_result = await db.execute(count_query)
            message_count = count_result.scalar() or 0

            responses.append(
                SessionResponse(
                    id=session.id,
                    user_id=session.user_id,
                    title=session.title,
                    description=session.description,
                    analysis_goal=session.analysis_goal,
                    message_count=message_count,
                    is_active=session.is_active,
                    created_at=session.created_at.replace(tzinfo=KST),
                    updated_at=session.updated_at.replace(tzinfo=KST),
                )
            )

        return responses, total_count

    @staticmethod
    async def update_session(
        db: AsyncSession,
        user_id: int,
        session_id: int,
        request: SessionUpdateRequest,
    ) -> SessionResponse | None:
        """
        Update session metadata (title, description).

        Args:
            db: AsyncSession database connection
            user_id: User ID (for authorization)
            session_id: Session ID to update
            request: SessionUpdateRequest with new data

        Returns:
            Updated SessionResponse or None if not found
        """
        query = select(Session).where(
            (Session.id == session_id) & (Session.user_id == user_id)
        )
        result = await db.execute(query)
        session = result.scalar_one_or_none()

        if not session:
            return None

        # Update fields if provided
        if request.title is not None:
            session.title = request.title
        if request.description is not None:
            session.description = request.description

        await db.commit()

        # Count messages
        count_query = select(func.count(ChatMessage.id)).where(
            ChatMessage.session_id == session_id
        )
        count_result = await db.execute(count_query)
        message_count = count_result.scalar() or 0

        return SessionResponse(
            id=session.id,
            user_id=session.user_id,
            title=session.title,
            description=session.description,
            analysis_goal=session.analysis_goal,
            message_count=message_count,
            is_active=session.is_active,
            created_at=session.created_at.replace(tzinfo=KST),
            updated_at=session.updated_at.replace(tzinfo=KST),
        )

    @staticmethod
    async def update_analysis_goal(
        db: AsyncSession,
        user_id: int,
        session_id: int,
        request: SessionUpdate,
    ) -> SessionResponse | None:
        """
        Update session analysis goal.

        Args:
            db: AsyncSession database connection
            user_id: User ID (for authorization)
            session_id: Session ID to update
            request: SessionUpdate with analysis_goal

        Returns:
            Updated SessionResponse or None if not found
        """
        query = select(Session).where(
            (Session.id == session_id) & (Session.user_id == user_id)
        )
        result = await db.execute(query)
        session = result.scalar_one_or_none()

        if not session:
            return None

        # Update analysis_goal if provided
        if request.analysis_goal is not None:
            session.analysis_goal = request.analysis_goal

        await db.commit()

        # Count messages
        count_query = select(func.count(ChatMessage.id)).where(
            ChatMessage.session_id == session_id
        )
        count_result = await db.execute(count_query)
        message_count = count_result.scalar() or 0

        return SessionResponse(
            id=session.id,
            user_id=session.user_id,
            title=session.title,
            description=session.description,
            analysis_goal=session.analysis_goal,
            message_count=message_count,
            is_active=session.is_active,
            created_at=session.created_at.replace(tzinfo=KST),
            updated_at=session.updated_at.replace(tzinfo=KST),
        )

    @staticmethod
    async def delete_session(
        db: AsyncSession,
        user_id: int,
        session_id: int,
    ) -> SessionDeleteResponse | None:
        """
        Delete a session and all its messages.

        Args:
            db: AsyncSession database connection
            user_id: User ID (for authorization)
            session_id: Session ID to delete

        Returns:
            SessionDeleteResponse with deletion details or None if not found
        """
        # Get session (authorization check)
        query = select(Session).where(
            (Session.id == session_id) & (Session.user_id == user_id)
        )
        result = await db.execute(query)
        session = result.scalar_one_or_none()

        if not session:
            return None

        # Count messages before deletion
        count_query = select(func.count(ChatMessage.id)).where(
            ChatMessage.session_id == session_id
        )
        count_result = await db.execute(count_query)
        message_count = count_result.scalar() or 0

        # Delete messages
        delete_msg_query = ChatMessage.__table__.delete().where(
            ChatMessage.session_id == session_id
        )
        await db.execute(delete_msg_query)

        # Delete session
        await db.delete(session)
        await db.commit()

        return SessionDeleteResponse(
            session_id=str(session_id),
            success=True,
            deleted_message_count=message_count,
            timestamp=datetime.now(KST),
        )

File: frontend/src/services/sessionService.js
import apiClient from './apiClient';
import config from '../config/config';

/**
 * Session Service
 * 세션 관련 API 호출을 관리합니다.
 */

const sessionService = {
  /**
   * 세션 목록 조회
   * @param {number} limit - 조회할 세션 수 (기본값: 100)
   * @param {number} offset - 오프셋 (기본값: 0)
   */
  async getSessions(limit = 100, offset = 0) {
    try {
      const response = await apiClient.get(
        `${config.api.endpoints.sessions.list}?limit=${limit}&offset=${offset}`
      );
      return response;
    } catch (error) {
      throw this.handleError(error);
    }
  },

  /**
   * 세션 생성
   * @param {Object} sessionData - 세션 데이터
   * @param {string} sessionData.title - 세션 제목
   * @param {string} sessionData.description - 세션 설명
   */
  async createSession(sessionData) {
    try {
      const response = await apiClient.post(config.api.endpoints.sessions.create, sessionData);
      return response;
    } catch (error) {
      throw this.handleError(error);
    }
  },

  /**
   * 세션 상세 조회
   * @param {number} sessionId - 세션 ID
   */
  async getSession(sessionId) {
    try {
      const response = await apiClient.get(config.api.endpoints.sessions.get(sessionId));
      return response;
    } catch (error) {
      throw this.handleError(error);
    }
  },

  /**
   * 세션 수정
   * @param {number} sessionId - 세션 ID
   * @param {Object} sessionData - 수정할 세션 데이터
   */
  async updateSession(sessionId, sessionData) {
    try {
      const response = await apiClient.put(
        config.api.endpoints.sessions.update(sessionId),
        sessionData
      );
      return response;
    } catch (error) {
      throw this.handleError(error);
    }
  },

  /**
   * 세션 삭제
   * @param {number} sessionId - 세션 ID
   */
  async deleteSession(sessionId) {
    try {
      await apiClient.delete(config.api.endpoints.sessions.delete(sessionId));
      return true;
    } catch (error) {
      throw this.handleError(error);
    }
  },

  /**
   * 세션 분석 목표 업데이트
   * @param {number} sessionId - 세션 ID
   * @param {string} analysisGoal - 분석 목표
   */
  async updateAnalysisGoal(sessionId, analysisGoal) {
    try {
      const response = await apiClient.put(
        `${config.api.endpoints.sessions.get(sessionId)}/goal`,
        { analysis_goal: analysisGoal }
      );
      return response;
    } catch (error) {
      throw this.handleError(error);
    }
  },

  /**
   * 에러 처리
   */
  handleError(error) {
    if (error.response) {
      const status = error.response.status;
      
      switch (status) {
        case 400:
          return new Error('잘못된 요청입니다.');
        case 401:
          return new Error('인증이 필요합니다. 다시 로그인해주세요.');
        case 403:
          return new Error('접근 권한이 없습니다.');
        case 404:
          return new Error('세션을 찾을 수 없습니다.');
        case 500:
          return new Error('서버 오류가 발생했습니다.');
        default:
          return new Error(error.response.data?.detail || '오류가 발생했습니다.');
      }
    }
    
    if (error.message === 'timeout') {
      return new Error('요청 시간이 초과되었습니다.');
    }
    
    return new Error('네트워크 오류가 발생했습니다.');
  },
};

export default sessionService;
